src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -57,37 +76,182 @@ pub mod x86_64 {
+        pub flags: crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags,
+        /// All registers to be modified under the sub-leaf.
+        pub modifiers: Vec<CpuidRegisterModifier>,
     }
 
     /// Wrapper type to containing x86_64 CPU config modifiers.
     #[derive(Debug, Deserialize, Eq, PartialEq)]
     pub struct X86_64CpuTemplate {
         /// Modifiers for CPUID configuration.
-        // pub cpuid_modifiers: Vec<RegisterBitMask>,
+        #[serde(default)]
+        pub cpuid_modifiers: Vec<CpuidLeafModifier>,
         /// Modifiers for model specific registers.
+        #[serde(default)]
         pub msr_modifiers: Vec<RegisterModifier>,
     }
 
+    /// Bit mask to adjust targeted bits to the value,
+    /// applied by the filter only.
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct BitMask {
+        /// Mask value is to be applied according
+        /// to what is allowed through by the filter.
+        #[serde(deserialize_with = "deserialize_u64_from_str")]
+        pub filter: u64,
+        /// Mask value to be applied.
+        #[serde(deserialize_with = "deserialize_u64_from_str")]
+        pub value: u64,
+    }
+
     /// Bit-mask based wrapper to apply
     /// changes to a given register's value
     #[derive(Debug, Deserialize, Eq, PartialEq)]
     pub struct RegisterModifier {
         /// Pointer of the location to be bit masked.
         #[serde(deserialize_with = "deserialize_u32_from_str")]
-        addr: u32,
+        pub addr: u32,
         /// Mask to be applied to register's value at the address
         /// provided.
-        mask: BitMask,
+        pub mask: BitMask,
+    }
+
+    /// CPU template with modifiers to be applied on
+    /// top of an existing configuration to generate
+    /// the guest configuration to be used.
+    pub fn create_guest_cpu_config(
+        template: &X86_64CpuTemplate,
+        host_config: &X86_64CpuConfiguration,
+    ) -> Result<X86_64CpuConfiguration, Error> {
+        let mut guest_msrs_map: HashMap<u32, u64> = HashMap::new();
+        let mut guest_cpuid_map: BTreeMap<CpuidKey, CpuidEntry>;
+
+        // Get the hash map of CPUID data
+        if host_config.cpuid.amd().is_some() {
+            guest_cpuid_map = host_config.cpuid.amd().unwrap().0.clone();
+        } else if host_config.cpuid.intel().is_some() {
+            guest_cpuid_map = host_config.cpuid.intel().unwrap().0.clone();
+        } else {
+            return Err(Error::Cpuid);
+        }
+
+        // Apply CPUID modifiers
+        for leaf_mod in &template.cpuid_modifiers {
+            let cpuid_key = CpuidKey {
+                leaf: leaf_mod.leaf,
+                subleaf: leaf_mod.subleaf,
+            };
+
+            let cpuid_entry_option = guest_cpuid_map.get(&cpuid_key);
+            let mut guest_cpuid_entry = if let Some(entry) = cpuid_entry_option {
+                entry.clone()
+            } else {
+                CpuidEntry::default()
```

Comment:
This aspect of the behaviour needs to be clear for sure. I have gone this route to be the most resilient to failure, but it might be more desirable to actually fail the start of the guest.